File: example_propagator.py
"""Use W3C Trace Context and Baggage propagators, which are the OpenTelemetry defaults."""
from opentelemetry import trace, propagate, baggage
from opentelemetry.sdk.trace import TracerProvider
# from opentelemetry.sdk.trace.export import BatchSpanProcessor
# from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter

trace.set_tracer_provider(TracerProvider())
tracer = trace.get_tracer(__name__)  # Inject baggage
context = baggage.set_baggage("user_id", "alice")

# Propagate context with span
with tracer.start_as_current_span("parent-span", context=context) as parent_span:
    print("Trace ID:", parent_span.get_span_context().trace_id)
    print("Baggage user_id:", baggage.get_baggage("user_id", context))

    # Simulate outbound request: inject context (with active span and baggage) into headers
    headers: dict[str, str] = {}
    current_context = trace.set_span_in_context(parent_span, context)
    propagate.inject(headers, context=current_context)
    print("Injected headers:", headers)

# Simulate incoming request: extract context from headers
new_context = propagate.extract(headers)
with tracer.start_as_current_span("child-span", context=new_context) as child_span:
    print("Child Trace ID:", child_span.get_span_context().trace_id)
    print("Baggage in child:", baggage.get_baggage("user_id", new_context))

File: example_propagator.js
const { NodeTracerProvider } = require('@opentelemetry/sdk-trace-node');
const { W3CTraceContextPropagator, W3CBaggagePropagator, CompositePropagator } = require('@opentelemetry/core');
const { context, trace, propagation } = require('@opentelemetry/api');

const provider = new NodeTracerProvider();
provider.register({
  propagator: new CompositePropagator({
    propagators: [
      new W3CTraceContextPropagator(),
      new W3CBaggagePropagator(),
    ],
  }),
});
const tracer = trace.getTracer('sample-node-service');
const bag = propagation.createBaggage({ user_id: { value: "alice" } });
let ctx = propagation.setBaggage(context.active(), bag);
const parentSpan = tracer.startSpan('parent-span', undefined, ctx);
console.log("Trace ID:", parentSpan.spanContext().traceId);
const carrier = {};
propagation.inject(trace.setSpan(ctx, parentSpan), carrier);
console.log("Injected headers:", carrier);
const extractedContext = propagation.extract(context.active(), carrier);
const childSpan = tracer.startSpan('child-span', undefined, extractedContext);
console.log("Child Trace ID:", childSpan.spanContext().traceId);
const extractedBaggage = propagation.getBaggage(extractedContext);
console.log("Baggage in child:", extractedBaggage.getEntry("user_id").value);
childSpan.end();  parentSpan.end();
